streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -303,15 +320,38 @@ private void shutdownTasksAndState(final boolean rethrowExceptions) {
         producer.flush();
         // Close all task state managers
         closeAllStateManagers(rethrowExceptions);
-        try {
-            // un-assign the change log partitions
-            restoreConsumer.assign(Collections.<TopicPartition>emptyList());
-        } catch (Exception e) {
-            log.error("{} Failed to un-assign change log partitions: ", logPrefix, e);
-            if (rethrowExceptions) {
-                throw e;
-            }
-        }
+        // remove the changelog partitions from restore consumer
+        unAssignChangeLogPartitions(rethrowExceptions);
+    }
+
+    /**
+     * Similar to shutdownTasksAndState, however does not close the task managers,
+     * in the hope that soon the tasks will be assigned again
+     * @param rethrowExceptions
+     */
+    private void suspendTasksAndState(final boolean rethrowExceptions) {
+        log.debug("{} suspendTasksAndState: suspending all active tasks [{}] and standby tasks [{}]", logPrefix,
+            activeTasks.keySet(), standbyTasks.keySet());
+
+        // Commit first as there may be cached records that have not been flushed yet.
+        commitOffsets(rethrowExceptions);
```

Comment:
We still need to "close" the topology by calling the `close()` call on the processor nodes following the DAG ordering to finish up any user-customizable computing logic, since if the tasks will not be resumed after the rebalance, we need to make sure that they are restarted from a clean state on the other client.

However, some `close()` function implementation maybe not-undoable when the tasks are resumed: think if users decide to close any other external resources in the `close()` call or simply look at `KeyValuePrinterProcessor`. 

So I think we cannot save the cost of closing the topology and re-initialize it: we still need to close the processor nodes, and then upon resuming we should call `init` on the processor nodes again: maybe we can extract the code

```
// initialize the task by initializing all its processor nodes in the topology
        log.info("{} Initializing processor nodes of the topology", logPrefix);
        for (ProcessorNode node : this.topology.processors()) {
            this.currNode = node;
            try {
                node.init(this.processorContext);
            } finally {
                this.currNode = null;
            }
        }
```

in `AbstractTask` with a separate `init` method?